src/main/java/com/spotify/dns/AbstractChangeNotifier.java

```diff
@@ -34,18 +36,25 @@
 
   private final AtomicReference<Listener<T>> listenerRef = new AtomicReference<Listener<T>>();
 
+  private final AtomicBoolean listenerNotified = new AtomicBoolean(false);
+
+  private ReentrantLock lock = new ReentrantLock();
```

Comment:
The idea was to hide the lock from other code to reduce risk of deadlocks being introduced by calling code or derived classes (say a change notifier subclass also synchronizing on "this"), but TBH java sync primitives are not my forte.